docs/proposals/partition-ring-multi-az-replication.md

```diff
@@ -0,0 +1,209 @@
+
+1. Hashes the series labels to get a hash value
+2. Finds the primary ingester (smallest token > hash value)
+3. When replication is enabled, selects additional replicas by moving clockwise around the ring
+4. Ensures replicas are distributed across different availability zones
+
+The issue arises when replication is enabled: each series in a request is hashed independently, causing each series to route to different groups of ingesters.
+
+```mermaid
+graph TD
+    A[Write Request] --> B[Distributor]
+    B --> C[Hash Series 1] --> D[Ingesters: 5,7,9]
+    B --> E[Hash Series 2] --> F[Ingesters: 5,3,10]
+    B --> G[Hash Series 3] --> H[Ingesters: 7,27,28]
+    B --> I[...] --> J[Different ingester sets<br/>for each series]
+```
+
+## Problem
+
+### Limited AZ Failure Tolerance with replication factor
+
+While the token ring effectively distributes load across the ingester fleet, the independent hashing and routing of each series creates an amplification effect where a single ingester failure can impact a large number of write requests.
+
+Consider a ring with 30 ingesters, each series gets distributed to three different ingesters:
+
+```
+Sample 1: {name="http_request_latency",api="/push", status="2xx"}
+         → Ingesters: ing-5, ing-7, ing-9
+Sample 2: {name="http_request_latency",api="/push", status="4xx"}
+         → Ingesters: ing-5, ing-3, ing-10
+Sample 3: {name="http_request_latency",api="/push", status="2xx"}
+         → Ingesters: ing-7, ing-27, ing-28
+...
+```
+If ingesters `ing-15` and `ing-18` (in different AZs) are offline, any request containing a series that needs to write to both these ingesters will fail completely:
+
+```
+Sample 15: {name="http_request_latency",api="/push", status="5xx"}
+          → Ingesters: ing-10, ing-15, ing-18  // Request fails
+```
+
+With requests increasing their batch size, the probability of request failure becomes critical in replicated deployments. Given two failed ingesters in different AZs, each individual series has a small chance of requiring both failed ingesters. However, as request batch sizes increase, the probability that at least one series in the batch will hash to both failed ingesters approaches certainty.
+
+**Note**: This problem specifically affects Cortex using replication. Replication as 1 are not impacted by this availability amplification issue.
+
+## Proposed Solution
+
+### Partition Ring Architecture
+
+A new Partition Ring is proposed where the ring is divided into partitions, with each partition containing a set of tokens and a group of ingesters. Ingesters are allocated to partitions based on their order in the zonal StatefulSet, ensuring that scaling operations align with StatefulSet's LIFO behavior. Each partition contains a number of ingesters equal to the replication factor, with exactly one ingester per availability zone.
```

Comment:
Considering your well described current behavior 

> 1. Hashes the series labels to get a hash value
> 2. Finds the primary ingester (smallest token > hash value)
> 3. When replication is enabled, selects additional replicas by moving clockwise around the ring
> 4. Ensures replicas are distributed across different availability zones

We will now find the next partition instead of the next ingester (2) using the same hash as before. Finding the 3 ingesters from the partition (3). Ingesters in partitions are already in different zones, so no need for 4.

Am I understanding it correctly?